app/src/main/java/protect/card_locker/importexport/StocardImporter.java

```diff
@@ -121,173 +161,183 @@ public ZIPData importZIP(ZipInputStream zipInputStream, final ZIPData zipData) t
-                    loyaltyCardHashMap = appendToHashMap(
-                            loyaltyCardHashMap,
-                            cardName,
-                            "backImage",
-                            ZipUtils.readImage(zipInputStream)
-                    );
+                    record.backImage = ZipUtils.readImage(zipInputStream);
+                } else if (!fileName.endsWith("/")) {
+                    Log.d(TAG, "Unknown or unused loyalty-cards file " + fileName + ", skipping...");
                 }
+            } else if (!fileName.endsWith("/")) {
+                Log.d(TAG, "Unknown or unused file " + fileName + ", skipping...");
             }
         }
 
-        return new ZIPData(loyaltyCardHashMap, providers);
+        return new ZIPData(cards, providers);
     }
 
-    public ImportedData importLoyaltyCardHashMap(Context context, final ZIPData zipData) {
+    public ImportedData importLoyaltyCardHashMap(Context context, final ZIPData zipData) throws FormatException {
         ImportedData importedData = new ImportedData(new ArrayList<>(), new HashMap<>());
         int tempID = 0;
 
-        for (Map<String, Object> loyaltyCardData : zipData.loyaltyCardHashMap.values()) {
-            String providerId = (String) loyaltyCardData.get("_providerId");
+        List<String> cardKeys = new ArrayList<>(zipData.cards.keySet());
+        Collections.sort(cardKeys);
 
-            if (providerId == null) {
-                Log.d(TAG, "Missing providerId for card " + loyaltyCardData + ", ignoring...");
+        for (String key : cardKeys) {
+            StocardRecord record = zipData.cards.get(key);
+
+            if (record.providerId == null) {
+                Log.d(TAG, "Missing providerId for card " + record + ", ignoring...");
                 continue;
             }
 
-            HashMap<String, Object> providerData = zipData.providers.get(providerId);
+            if (record.cardId == null) {
+                throw new FormatException("No card ID listed, but is required");
+            }
+
+            StocardProvider provider = zipData.providers.get(record.providerId);
 
             // Read store from card, if not available (old export), fall back to providerData
-            String store;
-            if (loyaltyCardData.containsKey("store")) {
-                store = (String) loyaltyCardData.get("store");
-            } else {
-                store = providerData != null ? providerData.get("name").toString() : providerId;
+            String store = record.store != null ? record.store : provider != null ? provider.name : record.providerId;
+            String note = record.note != null ? record.note : "";
+            String barcodeTypeString = record.barcodeType != null ? record.barcodeType : provider != null ? provider.barcodeFormat : null;
+
+            if (record.label != null && !record.label.equals(store) && !record.label.equals(note)) {
+                String labelNote = "Label: " + record.label;
```

Comment:
Should this maybe be a translatable? I think "Label" may not be understandable in all languages

Reply:
I guess. We could also remove the prefix. Fidme doesn't use one and it stores various fields in the note.

Reply:
Removing the prefix sounds fine to me as well

Reply:
Done.